include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h

```diff
@@ -0,0 +1,681 @@
+{
+    using _base_type = __binhash_manager_base<_BinHash>;
+    using _extra_memory_type = typename _BinHash::_range_value_type;
+    //While this stays "unused" in this struct, __buffer is required to keep the sycl buffer alive until the kernel has
+    // been completed (waited on)
+    _BufferType __buffer;
+
+    __binhash_manager_with_buffer(_BinHash __bin_hash_, _BufferType __buffer_)
+        : _base_type(__bin_hash_), __buffer(__buffer_)
+    {
+    }
+
+    ::std::size_t
+    get_required_SLM_elements()
+    {
+        return this->__bin_hash.get_range().size();
+    }
+
+    void
+    require_access(sycl::handler& __cgh)
+    {
+        oneapi::dpl::__ranges::__require_access(__cgh, this->__bin_hash.get_range());
+    }
+
+    auto
+    get_device_copyable_binhash()
+    {
+        return _base_type::get_device_copyable_binhash();
+    }
+};
+
+template <typename _BinHash>
+auto
+__make_binhash_manager(_BinHash __bin_hash)
+{
+    return __binhash_manager_base(__bin_hash);
+}
+
+template <typename _Range>
+auto
+__make_binhash_manager(oneapi::dpl::__internal::__custom_range_binhash<_Range> __bin_hash)
+{
+    auto __range_to_upg = __bin_hash.get_range();
+    auto __keep_boundaries =
+        oneapi::dpl::__ranges::__get_sycl_range<oneapi::dpl::__par_backend_hetero::access_mode::read,
+                                                decltype(__range_to_upg.begin())>();
+    auto __buffer = __keep_boundaries(__range_to_upg.begin(), __range_to_upg.end());
+    return __binhash_manager_with_buffer(oneapi::dpl::__internal::__custom_range_binhash(__buffer.all_view()),
+                                         __buffer);
+}
+
+template <typename... _Name>
+class __histo_kernel_register_local_red;
+
+template <typename... _Name>
+class __histo_kernel_local_atomics;
+
+template <typename... _Name>
+class __histo_kernel_private_glocal_atomics;
+
+template <typename _HistAccessor, typename _OffsetT, typename _Size>
+void
+__clear_wglocal_histograms(const _HistAccessor& __local_histogram, const _OffsetT& __offset, _Size __num_bins,
+                           const sycl::nd_item<1>& __self_item)
+{
+    ::std::uint32_t __gSize = __self_item.get_local_range()[0];
+    ::std::uint32_t __self_lidx = __self_item.get_local_id(0);
+    ::std::uint8_t __factor = oneapi::dpl::__internal::__dpl_ceiling_div(__num_bins, __gSize);
+    ::std::uint8_t __k = 0;
+
+    for (; __k < __factor - 1; ++__k)
+    {
+        __local_histogram[__offset + __gSize * __k + __self_lidx] = 0;
+    }
+    // residual
+    if (__gSize * __k + __self_lidx < __num_bins)
+    {
+        __local_histogram[__offset + __gSize * __k + __self_lidx] = 0;
+    }
+    __dpl_sycl::__group_barrier(__self_item);
+}
+
+template <typename _BinIdxType, typename _Range, typename _HistReg, typename _BinFunc>
+void
+__accum_local_register_iter(_Range&& __input, const ::std::size_t& __index, _HistReg* __histogram, _BinFunc __func)
```

Comment:
Thanks, I've taken your suggestions for 1,2,3.  I would prefer to leave it as a separate helper function so it can be reused rather than repeated.  It also allows us to mirror the structure of the other kernels.